`crux-llvm`: Fix compilation error on Windows

`crux-llvm` currently fails to compile on Windows, as seen in [this CI job](https://github.com/GaloisInc/crucible/runs/6113020921?check_suite_focus=true#step:10:1334):

```
[1 of 2] Compiling RealMain         ( exe\windows\RealMain.hs, D:\\a\crucible\crucible\dist-newstyle\build\x86_64-windows\ghc-9.0.2\crux-llvm-0.5.0.99\x\crux-llvm\build\crux-llvm\crux-llvm-tmp\RealMain.o )

exe\windows\RealMain.hs:7:29: error:
Error:     • Couldn't match expected type: Maybe
                                      Crux.Config.Common.OutputOptions
                                    -> Crux.Log.OutputConfig CruxLLVMMain.CruxLLVMLogging
                  with actual type: IO
                                      (Maybe Crux.Config.Common.OutputOptions
                                       -> Crux.Log.OutputConfig CruxLLVMMain.CruxLLVMLogging)
    • In the first argument of ‘mainWithOutputConfig’, namely
        ‘defaultOutputConfig’
      In the first argument of ‘(>>=)’, namely
        ‘mainWithOutputConfig defaultOutputConfig’
      In the expression:
        mainWithOutputConfig defaultOutputConfig >>= exitWith
  |
7 | main = mainWithOutputConfig defaultOutputConfig >>= exitWith

  |                             ^^^^^^^^^^^^^^^^^^^
```

This was most likely introduced in a76d80daf645fa5935aabf62922f81ddd1485538 (but not caught by CI at the time).

--- a/.github/ci.sh
+++ b/.github/ci.sh
@@ -42,7 +42,7 @@ configure() {
   ghc_ver="$(ghc --numeric-version)"
   cp cabal.GHC-"$ghc_ver".config cabal.project.freeze
   cabal v2-update
-  cabal v2-configure -j2 --minimize-conflict-set
+  cabal v2-configure "$@" -j2 --enable-tests --minimize-conflict-set
   #tee -a cabal.project > /dev/null < cabal.project.ci
 }
 

--- a/.github/workflows/README.md
+++ b/.github/workflows/README.md
@@ -9,10 +9,11 @@ conventions when picking with operating systems to test:
 * In addition, we also test the previous LTS release of Ubuntu on the
   `crux-llvm-build.yml` workflow, but only using the latest supported version
   of GHC.
-* For each project, we test macOS only using the latest
-  supported version of GHC. This is because the turnaround time for macOS
-  builders on GitHub Actions is longer, so we try to avoid
-  bottlenecking CI workflows on macOS builds.
+* For each project, we test macOS and Windows only using the latest
+  supported version of GHC. This is because the turnaround time for macOS and
+  Windows builders on GitHub Actions is longer, so we try to avoid
+  bottlenecking CI workflows on these operating systems.
 
-  In the future, we wish to test Windows using the latest supported version
-  of GHC as well.
+  Note that we do not currently run LLVM-related tests on Windows as it is not
+  straightforward to obtain Windows builds of LLVM that include all of the
+  developer tools that we need, such as `llvm-link` and `llvm-as`.

--- a/.github/workflows/crucible-go-build.yml
+++ b/.github/workflows/crucible-go-build.yml
@@ -31,6 +31,8 @@ jobs:
         include:
           - os: macos-10.15
             ghc: 9.0.2
+          - os: windows-latest
+            ghc: 9.0.2
     name: crucible-go - GHC v${{ matrix.ghc }} - ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
@@ -90,14 +92,9 @@ jobs:
           setup_bin() { echo setup.${{ matrix.ghc }}; }
           with_ghc()  { $NS -c ${@}; }
           (cd crucible-go; with_ghc ghc -o $(setup_bin) $(setup_src) && ./$(setup_bin) clean)
-      - run: cabal update
-      - run: cabal configure crux-go --enable-tests
-      - name: Linux build
-        if: runner.os == 'Linux'
-        run: cabal build exe:crux-go
-      - name: MacOS build
-        if: runner.os == 'macOS'
-        # has a clean and retry in case the linking cache is bad (seems to happen frequently with macos).
+      - name: Configure
         shell: bash
-        run: |
-          cabal build exe:crux-go || { cabal clean; cabal build exe:crux-go; }
+        run: .github/ci.sh configure crux-go
+      - name: Build
+        shell: bash
+        run: .github/ci.sh build exe:crux-go

--- a/.github/workflows/crucible-jvm-build.yml
+++ b/.github/workflows/crucible-jvm-build.yml
@@ -31,6 +31,8 @@ jobs:
         include:
           - os: macos-10.15
             ghc: 9.0.2
+          - os: windows-latest
+            ghc: 9.0.2
     name: crucible-jvm - GHC v${{ matrix.ghc }} - ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
@@ -90,14 +92,9 @@ jobs:
           setup_bin() { echo setup.${{ matrix.ghc }}; }
           with_ghc()  { $NS -c ${@}; }
           (cd crucible-jvm; with_ghc ghc -o $(setup_bin) $(setup_src) && ./$(setup_bin) clean)
-      - run: cabal update
-      - run: cabal configure crucible-jvm --enable-tests
-      - name: Linux build
-        if: runner.os == 'Linux'
-        run: cabal build exe:crucible-jvm
-      - name: MacOS build
-        if: runner.os == 'macOS'
-        # has a clean and retry in case the linking cache is bad (seems to happen frequently with macos).
+      - name: Configure
         shell: bash
-        run: |
-          cabal build exe:crucible-jvm || { cabal clean; cabal build exe:crucible-jvm; }
+        run: .github/ci.sh configure crucible-jvm
+      - name: Build
+        shell: bash
+        run: .github/ci.sh build exe:crucible-jvm

--- a/.github/workflows/crucible-wasm-build.yml
+++ b/.github/workflows/crucible-wasm-build.yml
@@ -31,6 +31,8 @@ jobs:
         include:
           - os: macos-10.15
             ghc: 9.0.2
+          - os: windows-latest
+            ghc: 9.0.2
     name: crucible-wasm - GHC v${{ matrix.ghc }} - ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
@@ -90,14 +92,9 @@ jobs:
           setup_bin() { echo setup.${{ matrix.ghc }}; }
           with_ghc()  { $NS -c ${@}; }
           (cd crucible-wasm; with_ghc ghc -o $(setup_bin) $(setup_src) && ./$(setup_bin) clean)
-      - run: cabal update
-      - run: cabal configure crucible-wasm --enable-tests
-      - name: Linux build
-        if: runner.os == 'Linux'
-        run: cabal build exe:crucible-wasm
-      - name: MacOS build
-        if: runner.os == 'macOS'
-        # has a clean and retry in case the linking cache is bad (seems to happen frequently with macos).
+      - name: Configure
         shell: bash
-        run: |
-          cabal build exe:crucible-wasm || { cabal clean; cabal build exe:crucible-wasm; }
+        run: .github/ci.sh configure crucible-wasm
+      - name: Build
+        shell: bash
+        run: .github/ci.sh build exe:crucible-wasm

--- a/.github/workflows/crux-llvm-build.yml
+++ b/.github/workflows/crux-llvm-build.yml
@@ -50,6 +50,8 @@ jobs:
             ghc: 9.0.2
           - os: macos-10.15
             ghc: 9.0.2
+          - os: windows-latest
+            ghc: 9.0.2
     name: crux-llvm - GHC v${{ matrix.ghc }} - ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
@@ -122,15 +124,12 @@ jobs:
         run: |
           .github/ci.sh build exe:crux-llvm
           .github/ci.sh build exe:crux-llvm-for-ide
+          .github/ci.sh build exe:crux-llvm-svcomp
           .github/ci.sh build exe:uc-crux-llvm
 
       - shell: bash
-        name: Test crucible (Linux)
+        name: Test crucible
         run: .github/ci.sh test crucible
-        if: runner.os == 'Linux'
-        env:
-          LLVM_LINK: "llvm-link-10"
-          CLANG: "clang-10"
 
       - shell: bash
         name: Test crucible-symio (Linux)
@@ -182,14 +181,6 @@ jobs:
           restore-keys: llvm-11.0.0
 
       - shell: bash
-        name: Test crucible (macOS)
-        run: .github/ci.sh test crucible
-        if: runner.os == 'macOS'
-        env:
-          LLVM_LINK: "${{ github.workspace }}/llvm/bin/llvm-link"
-          CLANG: "${{ github.workspace }}/llvm/bin/clang"
-
-      - shell: bash
         name: Test crucible-llvm (macOS)
         run: .github/ci.sh test crucible-llvm
         if: runner.os == 'macOS'
@@ -205,43 +196,6 @@ jobs:
         env:
           LLVM_LINK: "${{ github.workspace }}/llvm/bin/llvm-link"
           CLANG: "${{ github.workspace }}/llvm/bin/clang-withIncl"
-
-      - shell: bash
-        name: Test crucible (Windows)
-        run: .github/ci.sh test crucible
-        if: runner.os == 'Windows'
-        env:
-          LLVM_LINK: "/c/Program Files/LLVM/bin/llvm-link"
-          CLANG: "/c/Program Files/LLVM/bin/clang"
-
-      - shell: bash
-        name: Test crucible-llvm (Windows)
-        run: .github/ci.sh test crucible-llvm
-        if: runner.os == 'Windows'
-        env:
-          LLVM_LINK: "/c/Program Files/LLVM/bin/llvm-link"
-          LLVM_AS: "/c/Program Files/LLVM/bin/llvm-as"
-          CLANG: "/c/Program Files/LLVM/bin/clang"
-
-      - shell: bash
-        name: Test crux-llvm (Windows)
-        run: .github/ci.sh test crux-llvm
-        if: runner.os == 'Windows'
-        env:
-          LLVM_LINK: "/c/Program Files/LLVM/bin/llvm-link"
-          CLANG: "/c/Program Files/LLVM/bin/clang"
-
-      - shell: bash
-        name: Test uc-crux-llvm (Windows)
-        run: .github/ci.sh test uc-crux-llvm
-        if: runner.os == 'Windows'
-        env:
-          LLVM_LINK: "/c/Program Files/LLVM/bin/llvm-link"
-          CLANG: "/c/Program Files/LLVM/bin/clang"
-
-      - shell: bash
-        run: .github/ci.sh build exe:crux-llvm-svcomp
-        if: runner.os != 'Windows'
 
       - shell: bash
         run: .github/ci.sh bundle_crux_llvm_files

--- a/cabal.GHC-8.10.7.config
+++ b/cabal.GHC-8.10.7.config
@@ -270,7 +270,7 @@ constraints: any.Cabal ==3.2.1.0,
              any.unbounded-delays ==0.1.1.1,
              any.uniplate ==1.6.13,
              any.unix ==2.7.2.2,
-             any.unix-compat ==0.5.3,
+             any.unix-compat ==0.5.4,
              unix-compat -old-time,
              any.unliftio-core ==0.2.0.1,
              any.unordered-containers ==0.2.16.0,
@@ -298,4 +298,4 @@ constraints: any.Cabal ==3.2.1.0,
              any.zlib ==0.6.2.3,
              zlib -bundled-c-zlib -non-blocking-ffi -pkg-config,
              any.zlib-bindings ==0.1.1.5
-index-state: hackage.haskell.org 2022-01-06T13:38:03Z
+index-state: hackage.haskell.org 2022-04-21T14:15:08Z

--- a/cabal.GHC-8.8.4.config
+++ b/cabal.GHC-8.8.4.config
@@ -271,7 +271,7 @@ constraints: any.Cabal ==3.0.1.0,
              any.unbounded-delays ==0.1.1.1,
              any.uniplate ==1.6.13,
              any.unix ==2.7.2.2,
-             any.unix-compat ==0.5.3,
+             any.unix-compat ==0.5.4,
              unix-compat -old-time,
              any.unliftio-core ==0.2.0.1,
              any.unordered-containers ==0.2.16.0,
@@ -299,4 +299,4 @@ constraints: any.Cabal ==3.0.1.0,
              any.zlib ==0.6.2.3,
              zlib -bundled-c-zlib -non-blocking-ffi -pkg-config,
              any.zlib-bindings ==0.1.1.5
-index-state: hackage.haskell.org 2022-01-06T13:38:03Z
+index-state: hackage.haskell.org 2022-04-21T14:15:08Z

--- a/cabal.GHC-9.0.2.config
+++ b/cabal.GHC-9.0.2.config
@@ -270,7 +270,7 @@ constraints: any.Cabal ==3.4.1.0,
              any.unbounded-delays ==0.1.1.1,
              any.uniplate ==1.6.13,
              any.unix ==2.7.2.2,
-             any.unix-compat ==0.5.3,
+             any.unix-compat ==0.5.4,
              unix-compat -old-time,
              any.unliftio-core ==0.2.0.1,
              any.unordered-containers ==0.2.16.0,
@@ -298,4 +298,4 @@ constraints: any.Cabal ==3.4.1.0,
              any.zlib ==0.6.2.3,
              zlib -bundled-c-zlib -non-blocking-ffi -pkg-config,
              any.zlib-bindings ==0.1.1.5
-index-state: hackage.haskell.org 2022-01-06T13:38:03Z
+index-state: hackage.haskell.org 2022-04-21T14:15:08Z

--- a/crux-llvm/exe/windows/RealMain.hs
+++ b/crux-llvm/exe/windows/RealMain.hs
@@ -4,4 +4,6 @@ import System.Exit
 import CruxLLVMMain (mainWithOutputConfig, defaultOutputConfig)
 
 main :: IO ()
-main = mainWithOutputConfig defaultOutputConfig >>= exitWith
+main =
+  do ec <- mainWithOutputConfig =<< defaultOutputConfig
+     exitWith ec

--- a/crux-llvm/for-ide/windows/RealMain.hs
+++ b/crux-llvm/for-ide/windows/RealMain.hs
@@ -1,10 +1,14 @@
 module RealMain (makeMain) where
 
-import CruxLLVMMain (defaultOutputConfig, mainWithOutputConfig)
-import System.Exit
+import Crux.Log ( OutputConfig )
+import Crux.Config.Common (OutputOptions)
+import CruxLLVMMain ( CruxLLVMLogging, defaultOutputConfig )
+import System.Exit (ExitCode, exitWith)
 
 makeMain ::
-  ((Maybe CruxOptions -> OutputConfig CruxLLVMLogging) -> IO ExitCode) ->
+  ((Maybe OutputOptions -> OutputConfig CruxLLVMLogging) -> IO ExitCode) ->
   IO ()
 makeMain mainWithOutputConfig =
-  mainWithOutputConfig defaultOutputConfig >>= exitWith
+  do
+    ec <- mainWithOutputConfig =<< defaultOutputConfig
+    exitWith ec

--- a/uc-crux-llvm/exe/windows/RealMain.hs
+++ b/uc-crux-llvm/exe/windows/RealMain.hs
@@ -4,4 +4,7 @@ import System.Exit
 import UCCrux.LLVM.Main (defaultOutputConfig, mainWithOutputConfig)
 
 main :: IO ()
-main = mainWithOutputConfig defaultOutputConfig >>= exitWith
+main =
+  do
+    ec <- mainWithOutputConfig =<< defaultOutputConfig
+    exitWith ec

